src/test/java/com/squareup/kotlinpoet/CrossplatformTest.kt

```diff
@@ -0,0 +1,167 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.squareup.kotlinpoet
+
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.test.fail
+
+class CrossplatformTest {
+
+  @Test fun crossplatform() {
+    val headerTypeParam = TypeVariableName("V")
+    val headerType = "AtomicRef<V>"
+    val headerSpec = TypeSpec.headerClassBuilder(headerType)
+        .addModifiers(KModifier.INTERNAL)
+        .primaryConstructor(FunSpec.constructorBuilder()
+            .addParameter("value", headerTypeParam)
+            .build())
+        .addProperty(PropertySpec.builder("value", headerTypeParam).build())
+        .addFunction(FunSpec.builder("get")
+            .returns(headerTypeParam)
+            .build())
+        .addFunction(FunSpec.builder("set")
+            .addParameter("value", headerTypeParam)
+            .build())
+        .addFunction(FunSpec.builder("getAndSet")
+            .addParameter("value", headerTypeParam)
+            .returns(headerTypeParam)
+            .build())
+        .addFunction(FunSpec.builder("compareAndSet")
+            .addParameter("expect", headerTypeParam)
+            .addParameter("update", headerTypeParam)
+            .returns(Boolean::class)
+            .build())
+        .build()
+    val implName = ParameterizedTypeName.get(AtomicReference::class.asTypeName(), headerTypeParam)
+    val implSpec = TypeAliasSpec.builder(headerType, implName)
+        .addModifiers(KModifier.IMPL)
+        .build()
+    val fileSpec = FileSpec.builder("", "Test")
+        .addType(headerSpec)
+        .addTypeAlias(implSpec)
+        .build()
+
+    assertThat(fileSpec.toString()).isEqualTo("""
+      |import java.util.concurrent.atomic.AtomicReference
+      |import kotlin.Boolean
+      |
+      |header internal class AtomicRef<V>(value: V) {
+      |  val value: V
+      |
+      |  fun get(): V
+      |
+      |  fun set(value: V)
+      |
+      |  fun getAndSet(value: V): V
+      |
+      |  fun compareAndSet(expect: V, update: V): Boolean
+      |}
+      |
+      |impl typealias AtomicRef<V> = AtomicReference<V>
+      |""".trimMargin())
+  }
+
+  @Test fun headerWithSecondaryConstructors() {
+    val headerSpec = TypeSpec.headerClassBuilder("IoException")
+        .addModifiers(KModifier.OPEN)
+        .superclass(Exception::class)
+        .addFunction(FunSpec.constructorBuilder().build())
+        .addFunction(FunSpec.constructorBuilder()
+            .addParameter("message", String::class)
+            .build())
+        .build()
+    val fileSpec = FileSpec.builder("", "Test")
+        .addType(headerSpec)
+        .build()
+
+    assertThat(fileSpec.toString()).isEqualTo("""
+      |import java.lang.Exception
```

Comment:
This is weird! It should be `kotlin.Exception`. I'll have to look into that...

Reply:
`kotlin.Exception` is a type alias, so there's only `java.lang.Exception` at runtime. Guess the only way to fix this is to have some manual replacement logic.

Reply:
Yeah I mean I get why it's happening, but reflection-wise it's weird and it means there's weird implications for code generation now.

Reply:
I'd rather see it behave like `Any` which gets turned into `Object` but you don't see it.